bundle/jsky.app.ot/src/main/java/jsky/app/ot/viewer/SPElevationPlotPlugin.java

```diff
@@ -232,106 +212,122 @@ private void _setColorCode(String label) {
+
+    // Get all the legend items.
+    // Except for the special case of no color code, delegates all of the work to the createLegendItems... functions.
+    private LegendItemCollection getLegendItems() {
+        final ColorManager colorManager = ColorManager.instance;
+        colorManager.reset();
+
+        final LegendItemCollection lic;
+        switch (_colorCodeTrajectories) {
+            case _COLOR_CODE_OBS:
+                lic = createLegendItems(obs -> obs.getObservationIDAsString("unknown"));
+                break;
+
+            case _COLOR_CODE_PROG:
+                lic = createLegendItemsOpt(obs -> ImOption.apply(obs.getProgramID()).map(SPProgramID::stringValue));
+                break;
+
+            case _COLOR_CODE_BAND:
+                lic = createLegendItemsPre(tm -> {
+                    final LegendItemCollection plic = new LegendItemCollection();
+                    for (int i=0; i < 5; ++i) {
+                        final String band = (i == 0) ? "Default Band" : String.format("Band %d", i);
+                        final Paint p = (i == 0) ? Color.BLACK : colorManager.nextColor();
+                        tm.put(band, p);
+                        plic.add(new LegendItem(band, p));
+                    }
+                    return plic;
+                  }, obs -> {
+                    final SPProgram prog = (SPProgram) SPDB.get().lookupProgram(obs.getProgramKey()).getDataObject();
+                    final String bandStr = prog.getQueueBand();
+                    return (bandStr == null || bandStr.isEmpty()) ? "Default Band" : String.format("Band %s", bandStr);
+                });
+                break;
+
+            case _COLOR_CODE_PRIORITY:
+                lic = createLegendItemsPre(tm -> {
+                    final LegendItemCollection plic = new LegendItemCollection();
+                    for (final SPObservation.Priority prio: _PRIORITIES) {
+                        final Paint p = colorManager.nextColor();
+                        final String prioStr = String.format("%s Priority", prio.displayValue());
+                        tm.put(prioStr, p);
+                        plic.add(new LegendItem(prioStr, p));
+                    }
+                    return plic;
+                }, obs -> String.format("%s Priority", ((SPObservation) obs.getDataObject()).getPriority().displayValue()));
+                break;
+
+            case _COLOR_CODE_NONE:
+                // Special case: no legend, and all lines are drawn in black.
+                lic = new LegendItemCollection();
+                final Paint[] colors = new Paint[_selectedObservations.length];
+                java.util.Arrays.fill(colors, Color.BLACK);
+                _elevationPanel.setItemColors(colors);
+                break;
+
+            default:
+                // Shouldn't ever get here.
+                lic = null;
+                _elevationPanel.setItemColors(null);
+        }
+
+        return lic;
+    }
+
```

Comment:
Essentially, the pattern repeated here was to:
1. Possibly pre-create some legend items (e.g. based on priority or band); and then
2. Iterate over selected observations, creating legend items for them if necessary (depending on the menu options), and then assigning them colors based on some ID string extracted from the observation. (If this corresponded to an existing legend item, such as band, then the appropriate band color was chosen).

The pattern was repeated in each block of the switch / if, and the code was quite terrible and, as Rob pointed out, array indexing was done in a way that was incorrect with regards to fetching the next color, so I extracted out the pattern into the `createLegendItem...` methods and now delegate to them with lambda functions to do (1) the preprocessing, and (2) the extraction of some kind of ID string from an observation.

Colors are assigned using a `ColorManager` so that this code is isolated from the rest and used consistently, whereas before we had things like `_COLORS[nextIdx++ % _COLORS.length]`, with the mod operation sometimes being forgotten.